src/porepy/numerics/ad/operators.py

```diff
@@ -1750,88 +1798,111 @@ def __repr__(self) -> str:
         elif self.prev_time:
-            s += "Evaluated at the previous time step.\n"
+            s += f"Evaluated at the previous time step {self.time_step_index}.\n"
 
         return s
 
 
 class MixedDimensionalVariable(Variable):
     """Ad representation of a collection of variables that individually live on separate
-    subdomains or interfaces, but treated jointly in the mixed-dimensional sense.
+    subdomains or interfaces, but represent the same quantity and are treated jointly in
+    the mixed-dimensional sense.
+
+    Note:
+        As of now, the wrapped fixed-dimensional variables must fulfill the following
+        assumptions:
 
-    Conversion of the variables into numerical value should be done with respect to the
-    state of an array; see :meth:`Operator.evaluate`. Therefore, the MergedVariable does
-    not implement the method :meth:`Operator.parse`.
+        1. They have the same name
+        2. They are at the same time step and iterate.
+        3. They are defined in different grids (no overlaps).
 
     Parameters:
-        variables: List of variables to be merged. Should all have the same name.
+        variables: List of variables to be merged.
+
+    Raises:
+        AssertionError: If one of the above assumptions is violated.
 
     """
 
     def __init__(self, variables: list[Variable]) -> None:
-        ### PUBLIC
 
-        self.sub_vars = variables
-        """List of sub-variables passed at instantiation, each defined on a separate
-        domain.
-
-        """
+        time_indices = []
+        iterate_indices = []
+        names = []
+        domains = []
+
+        for var in variables:
+            time_indices.append(var.time_step_index)
+            iterate_indices.append(var.iterate_index)
+            names.append(var.name)
+            domains.append(var.domain)
+
+        # check assumptions
+        if len(variables) > 0:
+            assert (
+                len(set(time_indices)) == 1
+            ), "Cannot create md-variable from variables at different time steps."
+            assert (
+                len(set(iterate_indices)) == 1
+            ), "Cannot create md-variable from variables at different iterates."
+            assert (
+                len(set(names)) == 1
+            ), "Cannot create md-variable from variables with different names."
+            assert len(set(domains)) == len(
+                domains
+            ), "Cannot create md-variable from variables with overlapping domains."
+        # Default values for empty md variable
+        else:
+            time_indices = [-1]
+            iterate_indices = [0]
+            names = ["empty_md_variable"]
 
-        self.id = next(Variable._ids)
-        """ID counter. Used to identify variables during operator parsing."""
+        ### PRIVATE
+        self._id = next(Variable._ids)
+        # NOTE private time step index is -1 if public time step index of atomic
+        # variables is None (current time)
```

Comment:
Why not use same (None or +1) in both cases?

Reply:
I sticked with the convention that the first previous time step is represented by the index 0. Should I change it?

Reply:
But don't you use None for variables and -1 for md variables?

Reply:
This issue is related to your other comment, https://github.com/pmgbergen/porepy/pull/1166#discussion_r1590249513